src/specfem2D/moving_sources_par.F90

```diff
@@ -0,0 +1,1374 @@
+            !   endif
+            !   call synchronize_all()
+            !   if (myrank == 1) then
+            !     print *,myrank," ispec_selected_source_moving :", ispec_selected_source_moving(i_source, it_l), &
+            !             " islice_selected_source(i_source) :",islice_selected_source(i_source), "reset :", reset
+            !   endif
+            !   call synchronize_all()
+            !   if (myrank == 2) then
+            !     print *,myrank," ispec_selected_source_moving :", ispec_selected_source_moving(i_source, it_l), &
+            !             " islice_selected_source(i_source) :",islice_selected_source(i_source), "reset :", reset
+            !   endif
+            !   call synchronize_all()
+            !   if (myrank == 3) then
+            !     print *,myrank," ispec_selected_source_moving :", ispec_selected_source_moving(i_source, it_l), &
+            !             " islice_selected_source(i_source) :",islice_selected_source(i_source), "reset :", reset
+            !     print *,""
+            !   endif
+            !   call synchronize_all()
+            ! endif
+
+          endif ! abs(source_time_function(i_source,it_l,i_stage_loc)) > TINYVAL
+        enddo  ! NSOURCES
+        ! synchronizes all processes
+        call synchronize_all()
+      enddo ! stage_time_scheme
+
+      ! counts sources in this process slice for this time step
+      nsources_local = 0
+      do i_source = 1, NSOURCES
+        if (myrank == islice_selected_source(i_source)) then
+          nsources_local = nsources_local + 1
+        endif
+      enddo
+      nsources_local_moving(it_l) = nsources_local
+
+    enddo ! NSTEP
+
+    if (writeMovingDatabases) then  ! Write moving databases to a binary file
+      call write_moving_databases(pathToMovingDatabase)
+    endif
+  else  ! Read the binary file file_exists .and. writeMovingDatabases
+    call read_moving_databases(pathToMovingDatabase)
+  endif
+
+  ! Beware:
+  ! When the source is moving we don't know where it is going: all the slices
+  ! need to know the source_time_function
+  ! If the source is not moving only the slice containing the source knows the source_time_function
+
+  ! Send variables to device, once and for all:
+  call prepare_moving_sources_cuda(Mesh_pointer, nsources_local_moving, NSOURCES, &
+                                  sourcearrays_moving, ispec_selected_source_moving, &
+                                  NSTEP, source_time_function)
+
+  ! Free useless memory
+  do i_source = 1,NSOURCES
+    if (pt_first_guess(i_source)%allocated) then
+      deallocate(pt_first_guess(i_source)%array)
+      pt_first_guess(i_source)%allocated = .false.
+    else
+      nullify(pt_first_guess(i_source)%array)
+    endif
+    pt_first_guess(i_source)%size = 0
+  enddo
+
+end subroutine init_moving_sources_GPU
+
+!
+!-----------------------------------------------------------------------------------------
+!
+
+  subroutine locate_source_moving(x_source,z_source, &
```

Comment:
not sure how fast/slow this routine is or if it slows down the performance of the code. in case it becomes a bottleneck, we could see if a kdtree search would help.

Reply:
I have worked a lot on this. It is now quite optimized (I don't loop over all the elements at each time step to locate the source, I first loop over the neighbors). For the moment it is ok I guess.